### ISSUE TITLE
Support chrome.sockets.* APIs

### ISSUE THREAD

**Issue by lidel** [ipfs-shipyard/ipfs-companion#664]
## About APIs 
Chromium OS exposes `chrome.sockets.*` for its apps:

- https://developer.chrome.com/apps/manifest/sockets#manifest
  - https://developer.chrome.com/apps/sockets_tcp
  - https://developer.chrome.com/apps/sockets_tcpServer
  - https://developer.chrome.com/apps/sockets_udp

We can access those APIs in Brave  by adding ipfs-companion (beta+stable) as `blessed_extension` in [brave-core/../_api_features.json](https://github.com/brave/brave-core/blob/bc836d540a9aef6c5371fad89f5da2f77f7a212e/common/extensions/api/_api_features.json#L61-L75) (suggested by @bbondy)

> See also my notes on bringing IPFS to Brave in 2019: https://github.com/brave/brave-browser/issues/819#issuecomment-456039555

## Potential use

-  Exposing HTTP Gateway to embedded js-ipfs node
  - Making js-ipfs basically as useful as go-ipfs for regular user
  - I imagine we want to detect Brave in [js-ipfs/.../config-browser.js#L8](https://github.com/ipfs/js-ipfs/blob/v0.34.1/src/core/runtime/config-browser.js#L8) 
- Local Discovery and UDP/TCP Transports
  (similar to @alanshaw's [libdweb experiments on Firefox](https://github.com/ipfs-shipyard/ipfs-companion/pull/553))

## Relevant libraries

Created for and used in webtorrent:
- https://github.com/feross/chrome-net follows Node.js [net](https://nodejs.org/api/net.html) (TCP) API
- https://github.com/feross/chrome-dgram follows Node.js [dgram](http://nodejs.org/api/dgram.html) (UDP) API
- Use with browserify: [webtorrent/package.json#L28-L30](https://github.com/webtorrent/webtorrent/blob/79eb836fa5beb528748f2eb7fd053c4fe1d23ecd/package.json#L28-L30)

## MVP: Exposing HTTP Gateway

Quick notes about first steps:
1. [x] Create Brave build that exposes `chrome.sockets.*` APIs to temporary extensions
   - Check it it works by default. If not, find Brave build instructions here: https://github.com/brave/brave-browser/wiki and whitelist rules are around [brave-core/../_api_features.json](https://github.com/brave/brave-core/blob/bc836d540a9aef6c5371fad89f5da2f77f7a212e/common/extensions/api/_api_features.json#L61-L75)
2. [x] Detect js-ipfs is running in Brave and enable Gateway by providing a valid multiaddr instead of empty string
   - [config-nodejs.js#L10](https://github.com/ipfs/js-ipfs/blob/v0.34.2/src/core/runtime/config-nodejs.js#L10) vs [config-browser.js#L8](https://github.com/ipfs/js-ipfs/blob/v0.34.2/src/core/runtime/config-browser.js#L8)
3. [x] Try building ipfs-companion/js-ipfs with `chrome-net` and `chrome-ngram` as replacements for Node APIs in browser contexts
4. [x] Start ipfs-companion in Brave, switch to embedded js-ipfs node 
    - did it start and open Gateway on  TCP port specified in step 2? is HTTP Gateway working as expected?
      - if yes, then we are lucky and this it! :)
      - if not, then we need to write additional code for exposing Gateway over `chrome.sockets.*`
5. Continue in https://github.com/ipfs-shipyard/ipfs-companion/issues/716

> cc https://github.com/ipfs-shipyard/ipfs-companion/issues/312

**Comment by lidel** [comment]
Baby steps :) With [some help](https://github.com/brave/brave-browser/issues/3102) I was able to confirm APIs are available in ipfs-companion background page after whitelisting extension. Notes below.

### How To Develop with Local (Unpacked) Extension

0. (prerequisite) Created `bundle:brave` build script that adds [`sockets` section to extension manifest](https://developer.chrome.com/apps/manifest/sockets#manifest) – this is [already in this branch](https://github.com/ipfs-shipyard/ipfs-companion/tree/feat/brave-build-with-chrome-sockets)

1. Built (`npm run build`) and started Brave (`npm start`). 

2. Enabled "Developer mode" and loaded an unpacked extension with  , it got assigned ID `paanflafljinkomjoheoobkeigpakegp`. 

3. Generated ID shortcut in format expected by Brave..

   ```console
   $ echo -n paanflafljinkomjoheoobkeigpakegp | openssl sha1 | tr '[:lower:]' '[:upper:]'
   (STDIN)= CEBF34BE446B77F39E2C535566F8C27F9A460B1E
   ```
 
4. ..and added it to whitelists for `sockets.*`  APIs in two files in `src/brave` ([brave-core](https://github.com/brave/brave-core/)):

```diff
--- a/common/extensions/api/_api_features.json
+++ b/common/extensions/api/_api_features.json
@@ -61,17 +61,17 @@
   "sockets.tcp": {
     "dependencies": ["manifest:sockets"],
     "contexts": ["blessed_extension"],
-    "whitelist": ["3D9518A72EB02667A773B69DBA9E72E0F4A37423"]
+    "whitelist": ["3D9518A72EB02667A773B69DBA9E72E0F4A37423","CEBF34BE446B77F39E2C535566F8C27F9A460B1E"]
   },
   "sockets.tcpServer": {
     "dependencies": ["manifest:sockets"],
     "contexts": ["blessed_extension"],
-    "whitelist": ["3D9518A72EB02667A773B69DBA9E72E0F4A37423"]
+    "whitelist": ["3D9518A72EB02667A773B69DBA9E72E0F4A37423","CEBF34BE446B77F39E2C535566F8C27F9A460B1E"]
   },
   "sockets.udp": {
     "dependencies": ["manifest:sockets"],
     "contexts": ["blessed_extension"],
-    "whitelist": ["3D9518A72EB02667A773B69DBA9E72E0F4A37423"]
+    "whitelist": ["3D9518A72EB02667A773B69DBA9E72E0F4A37423","CEBF34BE446B77F39E2C535566F8C27F9A460B1E"]
   },
   "braveRewards": {
     "channel": "stable",
diff --git a/common/extensions/api/_manifest_features.json b/common/extensions/api/_manifest_features.json
index fcacdb73..db024427 100644
--- a/common/extensions/api/_manifest_features.json
+++ b/common/extensions/api/_manifest_features.json
@@ -6,6 +6,6 @@
   "sockets": {
     "channel": "stable",
     "extension_types": ["extension", "platform_app"],
-    "whitelist": ["3D9518A72EB02667A773B69DBA9E72E0F4A37423"]
+    "whitelist": ["3D9518A72EB02667A773B69DBA9E72E0F4A37423","CEBF34BE446B77F39E2C535566F8C27F9A460B1E"]
   }
 }

```

5. Rebuilt Brave via `npm run build` 
6. Started it via `npm start`
7. Loaded extension again, it got the same ID (`paanflafljinkomjoheoobkeigpakegp`)
8. Opened Console for Companion's background page via "Inspect views" on `chrome://extensions`  and confirmed `chrome.sockets` API is present :tada: 
   > ![brave-socket-apis-2019-01-25--16-45-18](https://user-images.githubusercontent.com/157609/51770308-b226d600-20e5-11e9-90c8-1111f0597070.png)

## Next

Override js-ipfs config to make it start Gateway and set up ipfs-companion build to see if use of `chrome-net` and `chrome-ngram` is enough to get it working.

**Comment by lidel** [comment]
First checkpoint, `http` backed by `http-node` and `chrome-net` works from ipfs-companion:

```js
let http = require('http')
let server = http.createServer(function (req, res) {                                                                                                                                                                                        
  res.writeHead(200, { 'Content-Type': 'text/plain' })                                                                                                                                                                                      
  res.end('Hello from ipfs-companion exposing HTTP via chrome.sockets in Brave :-)\n')                                                                                                                                                      
})                                                                                                                                                                                                                                          
server.listen(9091, '127.0.0.1')
```

> ![ipfs-companion-brave-http-smoke-test-success-2019-01-28--15-32-02](https://user-images.githubusercontent.com/157609/51852215-e25cb780-2325-11e9-9bc7-51bd550ec3ca.png)

## Next

Getting js-ipfs working may take more than this, as `http-node` was created from Node 6.x and modern js-ipfs throws https://github.com/ipfs/js-ipfs-http-client/issues/840.

**Comment by lidel** [comment]
Second checkpoint: I took js-ipfs `0.35.0-pre.0` for a spin.  
It ships with [hapi.js](https://github.com/hapijs) v18, HTTP server used for exposing Gateway. 

> **Note:** see [changes in `feat/brave-build-with-chrome-sockets` branch](https://github.com/ipfs-shipyard/ipfs-companion/compare/feat/brave-build-with-chrome-sockets)

- [x] start raw `http` server (`http.createServer`)
- [x] start raw `hapi` server (`Hapi.Server`) (screenshot below)
  - after some troubleshooting I managed to start raw Hapi server but got internal error for all responses
- [x] return valid response from  `hapi` server running in browser extension
  - got it to work by replacing `url` with [iso-url](https://www.npmjs.com/package/iso-url) during webpack build


> ### Demo: hapi v18 (HTTP server) running in ipfs-companion
>  ![hapi-over-chrome-sockets](https://user-images.githubusercontent.com/157609/53490443-fe25bb80-3a93-11e9-8dcb-a35640052c65.png) 
> `TypeError: this._dht.on is not a function` comes from libp2p, see _Next_ below

## Next

- [ ]  start embedded js-ipfs with Gateway exposed by embedded Hapi server
  - right now `new Ipfs(..)` fails due to `TypeError: this._dht.on is not a function`, which is quite surprisingly not related to HTTP server (which should just work). It seems libp2p does not disable DHT fully when we polyfill `net`/`dgram` using `chrome.sockets`, which results in  `TypeError` around [js-libp2p/src/index.js#L404](https://github.com/libp2p/js-libp2p/blob/4ed5c039fc8583b62cb1e41284a025f5d8d258bc/src/index.js#L404)
- [ ] start js-ipfs-http-client
  - Does not work for now due to `TypeError: res.req.getHeaders is not a function` – `http` API mismatch triggered in [js-ipfs-http-client/src/utils/send-request.js#L41](https://github.com/ipfs/js-ipfs-http-client/blob/d40cb6c24120ba47615c754efb053c74bb7ac0ac/src/utils/send-request.js#L41)

**Comment by drbh** [comment]
@lidel great progress! How'd you get the Hapi route response to not throw an error? When I try to get a response from `:9092/` I get a 500 and the console says `TypeError: Uncaught error: Cannot read property '_replied' of null` 

Aside from that, what are the next steps for handling the roadblocks you noted above? 

If we completely disable DHT - should the embedded js-ipfs Gateway run?

**Comment by lidel** [comment]
@drbh I believe `_replied` error went away when I switched to new hapi/js-ipfs. Try: 
- fetching recent updates from `feat/brave-build-with-chrome-sockets` branch 
- rebulding Companion via `npm run dev-build bundle:brave`

As for remaining roadblocks: 
  - `this._dht.on is not a function` – I tried to disable DHT via config passed to `new Ipfs()`, but the error remained – not sure why. It is something I will be looking next (need to understand code around [js-libp2p/src/index.js#L404](https://github.com/libp2p/js-libp2p/blob/4ed5c039fc8583b62cb1e41284a025f5d8d258bc/src/index.js#L404) first)
- `res.req.getHeaders is not a function` – this is a low priority (we want to get js-ipfs online first), but my current guess is this will probably require update of [http-node](https://www.npmjs.com/package/http-node) (`http` api shim which is 3 years old)

**Comment by lidel** [comment]
[Changes in `feat/brave-build-with-chrome-sockets` branch](https://github.com/ipfs-shipyard/ipfs-companion/compare/feat/brave-build-with-chrome-sockets):

- After some debugging I managed to fix `this._dht.on is not a function` by disabling DHT in `options.libp2p.config` (instead of `options.config`):
```js
{
  config: {
    Addresses: {
      Swarm: [],
      API: '/ip4/127.0.0.1/tcp/5002',
      Gateway: '/ip4/127.0.0.1/tcp/9090'
    }
  },
  libp2p: {
    config: {
      dht: {
        enabled: false
      }
    }
  }
}
```
- Switched to js-ipfs `0.35.0-rc.0`
  - Starts fine, but does not start HTTP server with API/Gateway
    This is because an entry point is initialized only via `jsipfs daemon` in cli. 
    - We need to refactor [src/cli/commands/daemon.js](https://github.com/ipfs/js-ipfs/blob/6b789ee2958f6d1b6340d9de9ded049f8a34ab2a/src/cli/commands/daemon.js) and make it possible to start it programmatically in browser environment (WIP)

**Comment by lidel** [comment]
[Changes in `feat/brave-build-with-chrome-sockets` branch](https://github.com/ipfs-shipyard/ipfs-companion/compare/feat/brave-build-with-chrome-sockets):


After some refactoring in https://github.com/ipfs/js-ipfs/pull/1950,  https://github.com/pinojs/pino/pull/612 and some ah-hoc polyfills, I am stoked to present..

### **Embedded js-ipfs with self-hosted HTTP API and Gateway!** :tada: 

| Gateway: dir listing | API: `/api/v0/id` | 
| ---- | ---- |
| ![working-dir-listing-2019-03-20--00-07-32](https://user-images.githubusercontent.com/157609/54647831-5dc82300-4aa4-11e9-9b6a-32b6c5f36507.png) | ![working-api-port-2019-03-19--22-41-20](https://user-images.githubusercontent.com/157609/54647832-5dc82300-4aa4-11e9-8c5f-f1b361a6a053.png) |

\o/

### Known Issues

- If you have an old js-ipfs config without API and Gateway ports, go to _Preferences_ of the extension,   scroll to the bottom and click on ![2019-03-20--00-28-21](https://user-images.githubusercontent.com/157609/54648573-17c08e80-4aa7-11e9-9980-d38b7268e10c.png)

- Gateway: **only directory listing works ATM,** fetching raw file data returns some JSON artifacts instead (probably need to tweak some polyfils related to buffers/streams)
- "power button" from browser action menu  does not stop the js-ipfs, [it hangs in "stopping state"](https://user-images.githubusercontent.com/157609/54648157-708f2780-4aa5-11e9-83d0-29c3aae2dae4.png) (need to look into this)
- Switching to "external" ipfs node does not work (but I believe it will be an easy fix)

**Comment by lidel** [comment]
[Changes in `feat/brave-build-with-chrome-sockets` branch](https://github.com/ipfs-shipyard/ipfs-companion/compare/feat/brave-build-with-chrome-sockets):

## Solved the problem with fetching raw file data

Hapi's stream detection via `stream instanceof Stream` does not work correctly in browser context, especially when different polyfils are used and mixed together.
    
I created a patch to detect browserified streams in Chrome App environment, which  replaces `instanceof` check with feature-detection. This enables Hapi to consume stream-like objects, as long they match proper contract. 


**tl;dr** loading data from embedded gateway works:

![2019-04-02--20-43-46](https://user-images.githubusercontent.com/157609/55488993-620e4780-5631-11e9-9895-817e3519db2c.png)


## Fixed streaming of compressed payload
    
There is a WIP js-ipfs patch to do proper streaming and content-type sniffing. It simplifies code responsible for streaming response and makes the streaming actually work by telling the payload compression stream to flush its content on every read(). IIUC previous version was buffering entire thing in Hapi's compressor memory, so the payload was returned when entire thing was fetched.

## Known Issues

- `Uncaught Error: Cannot call write after a stream was destroyed` on refreshing already loaded payload
- content-type sniffing needs refinement, current version was ported from libdweb experiment
- Artifact created by `npm run dev-build bundle:generic`  is not compatible with regular Chrome yet. We need to switch to `chrome.sockets` feature-detection at runtime, so it can be installed from regular Chrome Store.

## Next

- Solve known issues
- Cleanup patches and create upstream PRs
- Update Companion UI to be aware of embedded Gateway 
- Merge and publish to Beta Channel

**Comment by lidel** [comment]
[Changes in `feat/brave-build-with-chrome-sockets` branch](https://github.com/ipfs-shipyard/ipfs-companion/compare/feat/brave-build-with-chrome-sockets):

## Companion UI is now aware of embedded Gateway

Embedded node is now the default if `chrome.sockets` are present. 
Both main menu and onboarding pages use it and all features are enabled:

![brave-1-2019-04-10--14-51-33](https://user-images.githubusercontent.com/157609/55883985-8f19a780-5ba7-11e9-803b-ce9b911827d5.png)


## Implemented feature-detection for `chrome.sockets`

We can now run the same code in Chrome and Brave, and more advanced embedded node will be enabled automatically when Brave runtime is detected.

This work included:

- Patched chrome-net: lidel/chrome-net@838ffde
  > chrome-net did not check if `chrome.sockets.*` actually exist. This caused problems with extensions built for Chromium-based browsers that tried to maintain the single codebase and do feature-detection at runtime.

- Feature detection for embedded js-ipfs:
  >   If `chrome.sockets` is available, embedded-brave.js will be used instead of regular embedded.js

- Refactored initial boot logic on clean install
  >  When extension started for the very first time, the node got restarted multiple times due to config updates/migrations happening in `lib/options.js`. 
  > 
  > I simplified boot process to the point no restarts are  triggered by config updates on the first run. JS on landing page is also tweaked to properly receive notification about new peers being available. Tests for relevant code paths are updated.

- Added new  `ipfsNodeType` named `embedded:chromesockets`
  > For now we do naive check if `chrome.sockets.tcp*` APIs are available and switch to `ipfsNodeType=embedded:chromesockets` if `true`.
  > User can customize configuration of embedded js-ipfs node via Preferences screen where `embedded:chromesockets` supersedes regular `embedded` on Brave:
  > 
  > ![2019-04-10-153104_1081x718_scrot](https://user-images.githubusercontent.com/157609/55882964-b1122a80-5ba5-11e9-9d9c-c233bedd0863.png)
  >  API and Gateway ports used by extension will follow configuration of embedded js-ipfs. 

## Made it play nice with go-ipfs and js-ipfs daemons

- Switching to External node works again, user can switch between embedded (default in Brave) and external IPFS nodes

- Notify user about socket errors:
  >   If port is already taken, a notification "port already in use" is displayed.
  >   We will most likely improve this soon (eg. by finding the next free port) but is good enough for now.

- Use custom ports by default
  > Embedded js-ipfs with `chrome.sockets` now listens on custom ports, removing the need of changing configuration if someone is already running go-ipfs or js-ipfs.

## Web UI works!

Web UI can be opened from Companion's main  menu and works as expected:

> ![2019-04-10-145206_1081x718_scrot](https://user-images.githubusercontent.com/157609/55883212-35fd4400-5ba6-11e9-82f5-f117f3e7d46f.png)

> ![2019-04-10-145231_1081x718_scrot](https://user-images.githubusercontent.com/157609/55883211-35fd4400-5ba6-11e9-9a28-a52bbe0750b3.png)

Sidenote: ipfs-webui connects to embedded js-ipfs over HTTP API, but we could make it work via `window.ipfs` by blessing Web UI's CID to work over it without scoping/sandboxing.


## Known Issues

Key blockers **in bold**

- [ ] **Content discovery is sometimes extremely slow** when preload nodes are under load and slow to respond
  - We may mitigate this by enabling relay and other existing discovery methods such as webrtc or stardust
  - This will just go away when we add p2p transport that can interop with go nodes
- [ ] **DNSLink websites** (e.g. `/ipns/docs.ipfs.io/`) are not supported by js-ipfs yet (tracked in https://github.com/ipfs/js-ipfs/issues/1918)
- [ ] DNS lookups switched to `ipfs.io/api/v0/dns` for DNS resolution until DNSLink support lands in js-ipfs and we get better understanding how to operate `chrome.sockets.udp*` API. 
- [ ] **HAMT-shareded directories,** eg. `/ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/wiki/` do not work with js-ipfs (tracked in https://github.com/ipfs/js-ipfs/issues/1963)

## Next

- [ ] Update to latest js-ipfs
- [ ] Submit remainingupstream patches to js-ipfs
- [ ] Create a work plan for known issues (some are being addresssed already)

**Comment by lidel** [comment]
Feature-detection for `chrome.sockets` works as expected, landed in [v2.8.1.778](https://github.com/ipfs-shipyard/ipfs-companion/releases/tag/v2.8.1.778) (Beta)

Remaining work around supercharging Embedded JS-IPFS in Brave is tracked in https://github.com/ipfs-shipyard/ipfs-companion/issues/716
